source/sigv4.c

```diff
@@ -793,6 +854,236 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerNum ].key.dataLen;
+
+        if( keyLen + 1 > buffRemaining )
+        {
+            sigV4Status = SigV4InsufficientMemory;
+        }
+        else
+        {
+            for( i = 0; i < keyLen; i++ )
+            {
+                if( !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
+                {
+                    *pBufLoc = tolower( canonicalRequest->pHeadersLoc[ headerNum ].key.pData[ i ] );
+                }
+                else
+                {
+                    *pBufLoc = canonicalRequest->pHeadersLoc[ headerNum ].key.pData[ i ];
+                }
+
+                pBufLoc++;
+            }
+
+            *pBufLoc = ';';
+            pBufLoc++;
+            buffRemaining -= ( keyLen + 1 );
+            canonicalRequest->pBufCur = pBufLoc;
+            canonicalRequest->bufRemaining = buffRemaining;
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t appendSignedHeaders( size_t headerCount,
+                                              uint32_t flags,
+                                              CanonicalContext_t * canonicalRequest )
+    {
+        size_t noOfHeaders = 0, keyLen = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        for( noOfHeaders = 0; noOfHeaders < headerCount; noOfHeaders++ )
+        {
+            assert( ( canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData ) != NULL );
+            sigV4Status = writeSignedHeaderToString( noOfHeaders, flags, canonicalRequest );
+
+            if( sigV4Status != SigV4Success )
+            {
+                break;
+            }
+        }
+
+        if( sigV4Status == SigV4Success )
+        {
+            canonicalRequest->pBufCur--;
+            *canonicalRequest->pBufCur = '\n';
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t writeCanonicalHeaderToString( size_t headerIndex,
+                                                       CanonicalContext_t * canonicalRequest )
+    {
+        size_t buffRemaining, keyLen = 0, valLen = 0, i = 0, trimValueLen = 0;
+        char * pBufLoc;
+        const char * value;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerNum ].key.dataLen;
+        valLen = canonicalRequest->pHeadersLoc[ headerNum ].value.dataLen;
+
+        if( keyLen + valLen + 2 > buffRemaining )
+        {
+            sigV4Status = SigV4InsufficientMemory;
```

Comment:
If the Header Value contains spaces that will be trimmed, then this logic can result in a false failure. Thus, it would be better to check for buffer space while actually copying the trimmed header value byte by byte. 
The logic I suggested in [this comment](https://github.com/aws/SigV4-for-AWS-IoT-embedded-sdk/pull/28#discussion_r682165080) cane be updated to also check for buffer space availability before copying the header value byte into the canonical request buffer:

```
if ( ! isLeadingSpace( value, valueLen, i ) && ! isTrailingSpace( value, valueLen, i ) && ! isFirsInSequentialSpaces( value, valueLen, i ) )
{ 
    if( bufferRemaining < 2 )
    {
        sigV4Status = SigV4InsufficientMemory; 
        break:
    }
    else
    {
           *pBufLoc = canonicalRequest->pHeadersLoc[ headerNum ].value.pData[ i ];
             pBufLoc++;
             buffRemaining--;
     }
}
```